browser/tor/tor_profile_service_impl.cc

```diff
@@ -46,42 +148,29 @@ void TorProfileServiceImpl::ReLaunchTor(const TorConfig& config) {
   tor_launcher_factory_->ReLaunchTorProcess(config);
 }
 
-void TorProfileServiceImpl::SetNewTorCircuitOnIOThread(
-    const scoped_refptr<net::URLRequestContextGetter>& getter,
-    std::string host) {
-  DCHECK_CURRENTLY_ON(BrowserThread::IO);
-  const TorConfig tor_config = tor_launcher_factory_->GetTorConfig();
-  if (tor_config.empty())
-    return;
-  auto* proxy_resolution_service =
-    getter->GetURLRequestContext()->proxy_resolution_service();
-  DCHECK(proxy_resolution_service);
-  TorProxyConfigService::TorSetProxy(proxy_resolution_service,
-                                     tor_config.proxy_string(),
-                                     host,
-                                     &tor_proxy_map_,
-                                     true);
-}
+void TorProfileServiceImpl::SetNewTorCircuit(WebContents* tab) {
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
+  // track the webcontents lifetime so we don't reload if it has already
+  // been destroyed
+  auto tracker = std::make_unique<NewTorCircuitTracker>(tab);
+  auto callback = base::BindOnce(&OnNewTorCircuit, std::move(tracker));
 
-void TorProfileServiceImpl::SetNewTorCircuit(const GURL& request_url,
-                                             const base::Closure& callback) {
-  std::string isolation_key = CircuitIsolationKey(request_url);
-  if (isolation_key.empty())
-    return;
+  const GURL& url = tab->GetURL();
+
+  proxy_config_service_->SetNewTorCircuit(url);
+
+  // Force lookup to erase the old circuit and also get a callback
+  // so we know when it is safe to reload the tab
   auto* storage_partition =
-    BrowserContext::GetStoragePartitionForSite(profile_, request_url, false);
-
-  net::URLRequestContextGetter* url_request_context_getter =
-    storage_partition->GetURLRequestContext();
-  DCHECK(url_request_context_getter);
-
-  base::PostTaskWithTraitsAndReply(
-      FROM_HERE, {BrowserThread::IO},
-      base::Bind(&TorProfileServiceImpl::SetNewTorCircuitOnIOThread,
-                 base::Unretained(this),
-                 base::WrapRefCounted(url_request_context_getter),
-                 isolation_key),
-    callback);
+      BrowserContext::GetStoragePartitionForSite(profile_, url, false);
+  if (!storage_partition) {
+    storage_partition =
+        content::BrowserContext::GetDefaultStoragePartition(profile_);
+  }
+  auto proxy_lookup_client_ptr =
+      TorProxyLookupClient::CreateTorProxyLookupClient(std::move(callback));
+  storage_partition->GetNetworkContext()->LookUpProxyForURL(
```

Comment:
I don't think we still need `TorProxyLookupClient` anymore which is for old method clear map in ProxyResolutionService. We can just `tracker->NewIdentityLoaded` and then when it refreshes, `ResolveProxy` will be triggered and we should have new circuit.

Reply:
I originally created it because there were multiple calls and I wanted to share the code so while it technically could be combined with the tracker, it doesn't hurt anything to keep them separate and might be useful later